File: linear_regression.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# =========================
# Load Dataset
# =========================
df = pd.read_csv("data.csv")

x = df["Crabs"].values
y = df["Clams"].values

# =========================
# Initialize Parameters
# =========================
m = 0.0
c = 0.0

learning_rate = 0.001
epochs = 50000

n = len(x)
cost_history = []

# =========================
# Training
# =========================
for epoch in range(epochs):

    # Predictions
    y_pred = m * x + c

    # Cost Function (MSE/2)
    cost = (1 / (2 * n)) * np.sum((y_pred - y) ** 2)
    cost_history.append(cost)

    # Gradients
    dm = (1 / n) * np.sum((y_pred - y) * x)
    dc = (1 / n) * np.sum(y_pred - y)

    # Update Parameters
    m = m - learning_rate * dm
    c = c - learning_rate * dc

    # Print Progress
    if epoch % 500 == 0:
        print(f"Epoch: {epoch}")
        print(f"Cost : {cost:.4f}")
        print(f"m    : {m:.4f}")
        print(f"c    : {c:.4f}")
        print("-" * 30)

# =========================
# Final Model
# =========================
print("\nTraining Complete")
print(f"Final Slope (m): {m:.4f}")
print(f"Final Intercept (c): {c:.4f}")

# =========================
# Test Prediction
# =========================
crabs = 3

prediction = m * crabs + c

print(f"\nPredicted Clams for {crabs} crabs = {prediction:.2f}")


# =========================
# Graph 1: Data + Best Fit Line
# =========================

plt.figure(figsize=(8,5))

plt.scatter(x, y, label="Actual Data")

x_line = np.linspace(min(x), max(x), 100)
y_line = m * x_line + c

plt.plot(x_line, y_line, label="Best Fit Line")

plt.xlabel("Crabs")
plt.ylabel("Clams")
plt.title("Linear Regression")
plt.legend()
plt.grid(True)

plt.show()

# Predictions
y_pred = m * x + c

# R² Score
ss_res = np.sum((y - y_pred) ** 2)

ss_tot = np.sum((y - np.mean(y)) ** 2)

r2 = 1 - (ss_res / ss_tot)

print(f"R² Score = {r2:.4f}")
print(cost_history)


# =========================
# Graph 2: Cost vs Epoch
# =========================

plt.figure(figsize=(8,5))

plt.plot(cost_history)

plt.xlabel("Epoch")
plt.ylabel("Cost (J)")
plt.title("Cost Function Convergence")
plt.grid(True)

plt.show()

File: ML Using C++/linear_regression.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <cmath>

using namespace std;

int main() {

    vector<double> x;
    vector<double> y;

    // =====================
    // Read CSV File
    // =====================
    ifstream file("data.csv");

    string line;

    getline(file, line); // Skip header

    while(getline(file, line)) {

        stringstream ss(line);

        string crab, clam;

        getline(ss, crab, ',');
        getline(ss, clam, ',');

        x.push_back(stod(crab));
        y.push_back(stod(clam));
    }

    file.close();

    // =====================
    // Initialize
    // =====================
    double m = 0.0;
    double c = 0.0;

    double learning_rate = 0.001;

    int epochs = 50000;

    int n = x.size();

    // =====================
    // Training
    // =====================
    for(int epoch = 0; epoch < epochs; epoch++) {

        double cost = 0;
        double dm = 0;
        double dc = 0;

        for(int i = 0; i < n; i++) {

            double prediction = m * x[i] + c;

            cost += pow(prediction - y[i], 2);

            dm += (prediction - y[i]) * x[i];

            dc += (prediction - y[i]);
        }

        cost = cost / (2 * n);

        dm = dm / n;
        dc = dc / n;

        // Update Parameters
        m = m - learning_rate * dm;
        c = c - learning_rate * dc;

        if(epoch % 5000 == 0) {

            cout << "Epoch: " << epoch << endl;
            cout << "Cost : " << cost << endl;
            cout << "m    : " << m << endl;
            cout << "c    : " << c << endl;
            cout << "------------------------" << endl;
        }
    }

    // =====================
    // Final Model
    // =====================
    cout << "\nTraining Complete\n";

    cout << "Final Slope (m): " << m << endl;
    cout << "Final Intercept (c): " << c << endl;

    // Prediction Example
    double crabs = 4;

    double prediction = m * crabs + c;

    cout << "\nPrediction for 3 crabs: "
         << prediction << endl;

    return 0;
}